File: backend/challenges/api_types.py
from ninja import Schema
from typing import List, Optional
from uuid import UUID


class CodingProblemSummary(Schema):
    id: UUID
    title: str
    slug: str
    difficulty: str
    topics: List[str]
    stats: Optional[dict] = None


class CodingProblemDetail(CodingProblemSummary):
    description: str
    examples: List[dict]
    constraints: List[str]
    starter_code: str
    test_cases: List[dict]


class CreateMockInterviewSchema(Schema):
    problem_id: Optional[UUID] = None
    difficulty: Optional[str] = None  # "easy" | "medium" | "hard" — used when problem_id is absent
    list_slug: Optional[str] = None
    duration_minutes: int = 35


class MockInterviewResponse(Schema):
    id: UUID
    conversation_id: UUID
    problem: CodingProblemDetail
    duration_minutes: int
    evaluation: Optional[dict] = None
    final_code: str = ""


class SubmitParams(Schema):
    code: str
    language: str = "python"


class ProblemAttemptOut(Schema):
    id: UUID
    verdict: str
    passed_count: int
    total_count: int
    execution_time_ms: Optional[int] = None
    submitted_at: str
    code: str = ""


class MyProgressOut(Schema):
    problem_id: UUID
    title: str
    difficulty: str
    topics: List[str]
    solved: bool
    attempts: int
    best_passed: int = 0
    best_total: int = 0


class SolutionOut(Schema):
    code: str = ""
    explanation: str = ""
    complexity: str = ""
    available: bool = False


class ProblemListOut(Schema):
    slug: str
    name: str
    description: str = ""
    problem_slugs: List[str] = []
    count: int = 0


class TutorChatIn(Schema):
    message: str = ""
    code: str = ""


class TutorHistoryOut(Schema):
    history: List[dict] = []


class TutorChatOut(TutorHistoryOut):
    reply: str

File: frontend/src/types/challenges/api_types.ts
import type { TestCaseInput } from "../execution/api_types";

export type Difficulty = "easy" | "medium" | "hard";

export interface ProblemStats {
  attempts: number;
  solved_by_users: number;
}

export interface CodingProblemSummary {
  id: string;
  title: string;
  slug: string;
  difficulty: Difficulty;
  topics: string[];
  stats: ProblemStats | null;
}

export interface ProblemExample {
  input: string;
  output: string;
  explanation?: string;
}

export interface CodingProblemDetail extends CodingProblemSummary {
  description: string;
  examples: ProblemExample[];
  constraints: string[];
  starter_code: string;
  test_cases: TestCaseInput[];
}

export interface ProblemAttempt {
  id: string;
  verdict: "passed" | "failed" | "error" | "timeout";
  passed_count: number;
  total_count: number;
  execution_time_ms: number | null;
  submitted_at: string;
  code: string;
}

export interface MyProgress {
  problem_id: string;
  title: string;
  difficulty: Difficulty;
  topics: string[];
  solved: boolean;
  attempts: number;
  best_passed: number;
  best_total: number;
}

export interface TutorTurn {
  role: "user" | "assistant";
  content: string;
}